juju/juju
# Update azure provider to use new Azure SDK

The Azure SDK has been rewritten by Microsoft and the existing ones deprecated. In particular, these packages
```
github.com/Azure/azure-sdk-for-go
github.com/Azure/go-autorest/autorest
```
are deprecated.

This PR migrates the Azure provider to the new SDK packages, located under `github.com/Azure/azure-sdk-for-go/sdk`.
Sadly, not everything can be migrated as the new SDK is not fully complete. The things we can't do yet:
- create service principals (for credential management)
- have transport and http request/response mocks for testing

The legacy stuff mentioned above has been moved so that it all lives together under the azureauth package. We had that already, but now it fully contains the deprecated stuff that can be replaced later.

For 2.9, we still need to support really old legacy unmanaged storage. We stopped creating new models with it as far back as 2.2 and it is removed in 3.0. Due to the issues testing with 2 SDKs, the tests for legacy storage are removed even those the code still nominally supports it (but in reality, nothing uses it anymore).

The changes essentially are to accommodate the new Azure resource model types, the new result iterators, and new error handling. A nice things about the new SDK is retries for transient errors and auto token refresh are internal to the SDK so no need to do that by hand ourselves. Associated tests also deleted.

## QA steps

juju bootstrap azure
juju add-model test
juju create-storage-pool des azure encrypted=true vault-name-prefix=mysecrets vault-user-id=27f86c31-bca0-4a37-b233-0e5030107696
juju add-machine 

use the azure portal to check that the disk encryption set, vault and keys are created and that the root disk of the machine uses these

juju deploy ubuntu -n 10

check that all units come up (tests deployment conflict handling)

juju deploy postgresql --storage pgdata=1G
juju status --storage
juju storage --format yaml

check detach/attach works too

juju detach-storage pgdata/0
juju attach-storage postgresql/0 pgdata/0

juju destroy-model test --destroy-storage
